mlem/contrib/sagemaker/meta.py

```diff
@@ -0,0 +1,484 @@
+    """Default ECR repository"""
+
+    @property
+    def role_name(self):
+        return f"arn:aws:iam::{self.account}:role/{self.role}"
+
+    @staticmethod
+    def _create_and_upload_model_arch(
+        session: sagemaker.Session,
+        model: MlemModel,
+        bucket: str,
+        model_arch_location: str,
+    ) -> str:
+        with tempfile.TemporaryDirectory() as dirname:
+            model.clone(os.path.join(dirname, "model", "model"))
+            arch_path = os.path.join(dirname, "arch", MODEL_TAR_FILENAME)
+            os.makedirs(os.path.dirname(arch_path))
+            with tarfile.open(arch_path, "w:gz") as tar:
+                path = os.path.join(dirname, "model")
+                for file in os.listdir(path):
+                    tar.add(os.path.join(path, file), arcname=file)
+
+            model_location = session.upload_data(
+                os.path.dirname(arch_path),
+                bucket=bucket,
+                key_prefix=posixpath.join(
+                    model_arch_location, model.meta_hash()
+                ),
+            )
+
+            return model_location
+
+    @staticmethod
+    def _delete_model_file(session: sagemaker.Session, model_path: str):
+        s3_client = session.boto_session.client("s3")
+        if model_path.startswith("s3://"):
+            model_path = model_path[len("s3://") :]
+        bucket, *paths = model_path.split("/")
+        model_path = posixpath.join(*paths, MODEL_TAR_FILENAME)
+        s3_client.delete_object(Bucket=bucket, Key=model_path)
+
+    def deploy(self, meta: SagemakerDeployment):
+        with meta.lock_state():
+            state: SagemakerDeployState = meta.get_state()
+            redeploy = meta.model_changed()
+            state.previous = state.previous or SagemakerDeployState()
+
+            session, aws_vars = self.get_session_and_aws_vars(state.region)
+            if state.region is None:
+                state.region = aws_vars.region
+                meta.update_state(state)
+
+            if not meta.use_prebuilt and (state.image_tag is None or redeploy):
+                self._build_image(meta, state, aws_vars)
+
+            if state.model_location is None or redeploy:
+                self._upload_model(meta, state, aws_vars, session)
+
+            if (
+                state.endpoint_name is None
+                or redeploy
+                or state.endpoint_model_hash is not None
+                and state.endpoint_model_hash != state.model_hash
+            ):
+                if state.endpoint_name is None:
+                    self._deploy_model(meta, state, aws_vars, session)
+                else:
+                    self._update_model(meta, state, aws_vars, session)
+
+    def _update_model(
+        self,
+        meta: SagemakerDeployment,
+        state: SagemakerDeployState,
+        aws_vars: AWSVars,
+        session: sagemaker.Session,
+    ):
+        assert state.model_location is not None  # TODO
```

Comment:
what is the #TODO here?

Reply:
I think it's about asserts that should not be used in produciton code. Here it is needed because i separated steps into multiple functions and mypy looses "track" of `state.model_location` is always not None here. So we should somehow tell mypy that it's the case (I went for assert and left a TODO to think of something better later)